### PR TITLE
Porting stock move backdating

### PR DESCRIPTION
**Pull request by archetipo**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
I'm not sure that with new quant approach this module is still valid. Have you tried?

**Comment by archetipo** [comment]
Hi @pedrobaeza 
yes i've tried.. why you are not sure?

**Comment by pedrobaeza** [comment]
Because quants are created always in the current date, so there will be an incongruency. It will also can create a paradox like the one of the fictional movies (how can you delivery things that don't exist because they have been already delivered in the future?)

**Comment by archetipo** [comment]
@pedrobaeza   ok Thanks

**Comment by eLBati** [comment]
@archetipo please see https://github.com/archetipo/stock-logistics-workflow/pull/1

File: stock_move_backdating/README.rst [hunk 1]
```diff
@@ -0,0 +1,45 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Allows back-dating of stock moves
+===========
```
Comment (tafaRU) [inline stock_move_backdating/README.rst]: It has to be to  “underlined” using equal signs.

**Comment by gurneyalex** [comment]
needs rebase

**Comment by archetipo** [comment]
Thanks @gurneyalex

File: .travis.yml [hunk 2]
```diff
@@ -1,10 +1,5 @@
-language: python
-
 sudo: false
-cache:
-  apt: true
-  directories:
-    - $HOME/.pip-cache/
+cache: pip
```
Comment (eLBati) [inline .travis.yml]: @archetipo probably travis is broken due this change. Why is it there?

Comment (blaggacao) [inline .travis.yml]: I wondered, too.

File: .travis.yml [hunk 3]
```diff
@@ -1,10 +1,5 @@
-language: python
```
Comment (blaggacao) [inline .travis.yml:1]: Ignorant question: Is this sensible?

Comment (eLBati) [inline .travis.yml:1]: MQT scripts are python: https://github.com/OCA/maintainer-quality-tools/tree/master/travis

In case of doubts, I would just follow the sample file
https://github.com/OCA/maintainer-quality-tools/blob/master/sample_files/.travis.yml

Comment (blaggacao) [inline .travis.yml:1]: so, your answer is no, it is not sensible to remove it. But for whatever reason it is removed, we don't know if @archetipo  has done it on purpose.

**Comment by archetipo** [comment]
@eLBati  @blaggacao Thnaks! Travis.yaml was fixed

File: stock_move_backdating/__openerp__.py [hunk 4]
```diff
@@ -19,23 +19,18 @@
...
+
 {
-    "name": "Stock Move Backdating",
-    "version": "1.0",
-    'author': ['Marco Dieckhoff, BREMSKERL', 'Agile Business Group'],
-    "category": "Stock Logistics",
+    'name': 'Stock Move Backdating',
+    'version': '1.0',
+    'category': 'Stock Logistics',
+    'author': 'Bremskerl, Odoo Community Association (OCA)',
```
Comment (eLBati) [inline stock_move_backdating/__openerp__.py]: Why did you change `author`?

Comment (archetipo) [inline stock_move_backdating/__openerp__.py]: author is here https://github.com/OCA/stock-logistics-workflow/pull/142/files#diff-8ebcf17df81968ac09b007cf5d372958R28

Comment (eLBati) [inline stock_move_backdating/__openerp__.py]: @archetipo that's not a good reason I think :bowtie:

**Comment by archetipo** [comment]
Thanks @eLBati  your remarks were fixed.

**Comment by eLBati** [comment]
@archetipo please also move the module to https://github.com/archetipo/stock-logistics-workflow/blob/porting_stock_move_backdating/README.md#available-addons

File: stock_move_backdating/wizard/stock_transfer_details.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    _defaults = {
+        'date': datetime.now(),
+    }
+
+    @api.one
+    @api.constrains('date')
+    def check_date(self):
+        now = datetime.now().strftime(
+            DEFAULT_SERVER_DATETIME_FORMAT)
```
Comment (gurneyalex) [inline stock_move_backdating/wizard/stock_transfer_details.py]: I think there is a TZ mismatch here, which will cause a bug:  `self.date` is in UTC, while `now` is in the local TZ of the server. I think you need to use `datetime.utcnow()` to fix this.

File: stock_move_backdating/wizard/stock_transfer_details.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+from openerp.exceptions import Warning
+from datetime import datetime
+from openerp.tools import DEFAULT_SERVER_DATETIME_FORMAT
+
+
+class StockTransferDetailsItems(models.TransientModel):
+    _inherit = 'stock.transfer_details_items'
+
+    _defaults = {
+        'date': datetime.now(),
```
Comment (gurneyalex) [inline stock_move_backdating/wizard/stock_transfer_details.py]: please change this to `fields.Datetime.now` (without parenthesis) (recommended function for default value)

File: stock_move_backdating/wizard/stock_transfer_details.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @api.one
+    @api.constrains('date')
+    def check_date(self):
+        now = datetime.now().strftime(
```
Comment (gurneyalex) [inline stock_move_backdating/wizard/stock_transfer_details.py]: please change the call to `strftime` to a call to `fields.Datetime.to_string`

File: stock_move_backdating/model/stock.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    @api.multi
+    def action_done(self):
+        # do actual processing
+        result = super(stock_move, self).action_done()
+        # overwrite date field where applicable
+        for move in self:
+            if move.linked_move_operation_ids:
+                operation = move.linked_move_operation_ids[0]
+                move.date = operation.operation_id.date
+                move.refresh()
```
Comment (gurneyalex) [inline stock_move_backdating/model/stock.py]: I don't think this call is necessary (and if it is not, it will probably badly harm the performances). Can you check?

File: stock_move_backdating/model/stock.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                if move.quant_ids:
+                    for quant in move.quant_ids:
+                        quant.sudo().in_date = move.date
```
Comment (gurneyalex) [inline stock_move_backdating/model/stock.py]: since you are writing the same date on all the quants, please do this in a single call to write, rather than in a loop.

File: stock_move_backdating/model/stock.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    _inherit = "stock.quant"
+
+    @api.model
+    def _create_account_move_line(
+        self, quants, move, credit_account_id, debit_account_id, journal_id
+    ):
+        res = super(stock_quant, self)._create_account_move_line(
+            quants, move, credit_account_id, debit_account_id, journal_id
+        )
+        for o2m_tuple in res:
```
Comment (gurneyalex) [inline stock_move_backdating/model/stock.py]: This method has no automated test (see https://coveralls.io/builds/4018455/source?filename=stock_move_backdating%2Fmodel%2Fstock.py)

AFAICT, `stock_quant._create_account_move_line` returns `None` (see https://github.com/odoo/odoo/blob/8.0/addons/stock_account/stock_account.py#L237) so I expect this code to crash with None not being iterable. 

Please add an automated test exercising the method and fix the crash :smile_cat:

**Comment by gurneyalex** [comment]
@archetipo code review found a :bug:

File: stock_move_backdating/wizard/stock_transfer_details.py (hunk 5 above)
Comment (archetipo) [inline stock_move_backdating/wizard/stock_transfer_details.py]: @gurneyalex  ok Thanks

**Comment by archetipo** [comment]
@gurneyalex  @eLBati  thanks for review, all remarks are been fixed

File: README.md [hunk 11]
```diff
@@ -36,6 +36,7 @@ addon | version | summary
 [stock_picking_reorder_lines](stock_picking_reorder_lines/) | 8.0.0.1.0 | Provide a new field on stock moves, allowing to manage the orders of moves in a picking.
 [stock_split_picking](stock_split_picking/) | 8.0.1.0.0 | Split a picking in two unconfirmed pickings
 [stock_transfer_split_multi](stock_transfer_split_multi/) | 8.0.1.0.0 | In the stock transfer wizard, you can split by multiple units
+[stock_move_backdating](stock_move_backdating/) | 8.0.0.1.0 | Allows back-dating of stock moves
```
Comment (tafaRU) [inline README.md]: It's not necessary, it's done automatically.

File: stock_move_backdating/__openerp__.py [hunk 12]
```diff
@@ -19,23 +19,19 @@
...
 ##############################################################################
+
+
 {
-    "name": "Stock Move Backdating",
-    "version": "1.0",
-    'author': ['Marco Dieckhoff, BREMSKERL', 'Agile Business Group'],
-    "category": "Stock Logistics",
+    'name': 'Stock Move Backdating',
+    'version': '8.0.0.1.0',
```
Comment (tafaRU) [inline stock_move_backdating/__openerp__.py]: It should be **8.0.1.0.0**

Comment (archetipo) [inline stock_move_backdating/__openerp__.py]: here seems be possible to use also 8.0.0.1.0
https://github.com/archetipo/stock-logistics-workflow/blob/porting_stock_move_backdating/README.md#available-addons

**Comment by eLBati** [comment]
@archetipo please see commits made in august: they have wrong committer, like

```
$ git cat-file commit 7b7039fc6f45958949a8c4d760262bdc1211a6aa
tree fbd6ddcb413a6d7cc9e60ef34d5e8582e25ce95f
parent 6667cdcfcb9bd45bf4303552efd25982e24efb93
author Alexandre Fayolle <alexandre.fayolle@camptocamp.com> 1437771965 +0200
committer archetipo <archetipo@localhost> 1445587634 +0200
```

Rebase should fix them

**Comment by archetipo** [comment]
@eLBati  ok thnaks!

**Comment by archetipo** [comment]
@eLBati  rebase done

File: stock_move_backdating/__init__.py [hunk 13]
```diff
@@ -3,7 +3,7 @@
 #
 #    Copyright (C) 2012 BREMSKERL-REIBBELAGWERKE EMMERLING GmbH & Co. KG
 #    Author Marco Dieckhoff
-#    Copyright (C) 2013 Agile Business Group sagl (<http://www.agilebg.com>)
+#    Copyright (C) 2013 Agile Business Group (<http://www.agilebg.com>)
```
Comment (tafaRU) [inline stock_move_backdating/__init__.py]: Please update copyright year, you should set as `2013-15`

File: stock_move_backdating/model/__init__.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    Copyright (C) 2012 BREMSKERL-REIBBELAGWERKE EMMERLING GmbH & Co. KG
+#    Author Marco Dieckhoff
+#    Copyright (C) 2013 Agile Business Group (<http://www.agilebg.com>)
```
Comment (tafaRU) [inline stock_move_backdating/model/__init__.py]: Same as above.

File: stock_move_backdating/model/stock.py [hunk 15]
```diff
@@ -0,0 +1,41 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    Copyright (C) 2012 BREMSKERL-REIBBELAGWERKE EMMERLING GmbH & Co. KG
+#    Author Marco Dieckhoff
+#    Copyright (C) 2013 Agile Business Group sagl (<http://www.agilebg.com>)
```
Comment (tafaRU) [inline stock_move_backdating/model/stock.py]: Same as above.

Comment (tafaRU) [inline stock_move_backdating/model/stock.py]: For consistency, please remove sagl at the end of the author name ;-)

File: stock_move_backdating/tests/__init__.py [hunk 16]
```diff
@@ -0,0 +1,24 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    Copyright (C) 2012 BREMSKERL-REIBBELAGWERKE EMMERLING GmbH & Co. KG
+#    Author Marco Dieckhoff
+#    Copyright (C) 2013 Agile Business Group  (<http://www.agilebg.com>)
```
Comment (tafaRU) [inline stock_move_backdating/tests/__init__.py]: Same as above.

File: stock_move_backdating/wizard/__init__.py [hunk 17]
```diff
@@ -1,7 +1,7 @@
 # -*- coding: utf-8 -*-
 ##############################################################################
 #
-#    Copyright (C) 2013 Agile Business Group sagl (<http://www.agilebg.com>)
+#    Copyright (C) 2013 Agile Business Group (<http://www.agilebg.com>)
```
Comment (tafaRU) [inline stock_move_backdating/wizard/__init__.py]: Same as above.

File: stock_move_backdating/__openerp__.py [hunk 18]
```diff
@@ -19,23 +19,19 @@
...
+    ' Odoo Community Association (OCA)',
     'website': 'www.bremskerl.com',
-    'license': 'AGPL-3',
-    "depends": ["stock"],
-    "summary": "Allows back-dating of stock moves",
-    "description": """This module allows to register old stock moves
-    (with date != now).
-    On stock moves, user can specify the "Actual Movement Date", that will be
-    used as movement date""",
+    'depends': ['stock_account'],
```
Comment (tafaRU) [inline stock_move_backdating/__openerp__.py:15]: Why did you change it?

**Comment by tafaRU** [comment]
@archetipo, could you also please use simple headers following OCA template
https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L1-L3 ?

Thank you in advance.

File: stock_move_backdating/__openerp__.py (hunk 18 above)
Comment (archetipo) [inline stock_move_backdating/__openerp__.py:15]: @tafaRU 
because from version 8.0 stock.move is not  in stock but stock.account

**Comment by archetipo** [comment]
@tafaRU 
thanks for review, done

File: stock_move_backdating/__openerp__.py (hunk 18 above)
Comment (pedrobaeza) [inline stock_move_backdating/__openerp__.py:15]: stock.move is defined in _stock_. Module _stock_account_ is a link for invoicing pickings only.

Comment (archetipo) [inline stock_move_backdating/__openerp__.py:15]: It says Pedro , my mistake , the change need to import both stock that account , with one module useful to have all dependencies met , and then have the link with the billing . Sorry but I wrote the module 4 months ago ...

**Comment by lepistone** [comment]
:+1: thanks!

File: stock_move_backdating/tests/test_stock_move_backdating.py [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+    def setUp(self):
+        super(TestStockMoveBackdating, self).setUp()
+
+        self.demo_user = self.env.ref('base.user_demo')
+        self.env.uid = self.demo_user.id
+        self.move_model = self.env["stock.move"]
+        self.picking_model = self.env["stock.picking"]
+        self.product_model = self.env["product.product"]
+        self.location_model = self.env["stock.location"]
+        self.stock_transfer_details = self.registry("stock.transfer_details")
```
Comment (tafaRU) [inline stock_move_backdating/tests/test_stock_move_backdating.py]: Here, too, it would be better to use `self.env` instead of `self.registry`.

**Comment by archetipo** [comment]
@tafaRU  thanks done

**Comment by tafaRU** [comment]
@archetipo, some Travis build jobs failed, could you please check them?

**Comment by archetipo** [comment]
@tafaRU  thanks done

**Comment by eLBati** [comment]
:+1: thanks

**Comment by tafaRU** [comment]
@archetipo, please see https://github.com/archetipo/stock-logistics-workflow/pull/2

**Comment by tafaRU** [comment]
:+1:
